### DISCUSSION TITLE
Should we add post-synthesis manifest validation

### DISCUSSION THREAD

**Discussion by iliapolo** [cdk8s-team/cdk8s#936] [General]
Ability to define manifest validators to be executed after synthesis
There are plenty of pitfalls users may fall into when defining their manifests. Validation tools exist that will help users avoid those. For example:

- https://github.com/datreeio/datree
- https://github.com/zegl/kube-score
- https://github.com/FairwindsOps/polaris
- https://www.checkov.io/
- https://www.aquasec.com/
- https://docs.snyk.io/
- https://docs.paloaltonetworks.com/prisma/prisma-cloud/prisma-cloud-admin-compute/install/install_kubernetes

Since cdk8s allows us to run code during manifest generation, we can integrate these tools into the process, without users having to invoke them explicitly. Our init templates can offer these validations by default, and allow users to opt out, or vise versa. 

Add a key in `cdk8s.yaml` to describe what validation should take place:

```yaml
app: ts-node main.ts
validations: 
  - datree test
```

### Open questions

- Should we be doing this? why and why not?
- Should these only happen when synthesizing with the CLI? Maybe in-process as well?
- What configuration should be exposed? the entire validation command? the validator type?

**Comment by shimont** [comment]
I think that this is a good approach, the "[State of Kubernetes Security Report](https://www.redhat.com/rhdc/managed-files/cl-state-kubernetes-security-report-ebook-f29117-202106-en_0.pdf)" shows that most incidents are caused by misconfiguration (page 4). By adding this functionality, cdk8s can help users avoid those pitfalls.

**Reply by reitzig** [reply]
FWIW, my gut feeling matches the UNIX philosophy here: don't overcomplicated this tool, let it do one thing well. Chaining validators behind it (as part of CI pipelines, before deployment!) _should_ be easy -- do those accept a set of YAML files on STDIN, for instance?

**Reply by iliapolo** [reply]
@reitzig we were debating this quite a bit internally. Eventually we considered that, while chaining validators is **technically** easy, it can be procedurally complicated in large organizations. And streamlining this process within cdk8s can really help. 

Have a look at our [RFC](https://github.com/cdk8s-team/cdk8s/blob/epolon/rfc-manifest-validation/rfc/manifest-validation.md) for a comprehensive rundown of our thought process. 

THanks!

**Comment by eyarz** [comment]
validation tools to prevent pitfalls is great, but because with every new K8s release, there are new properties that are deprecated, I would also consider incorporating a tool to validate the K8s schema:
https://opensource.com/article/21/7/kubernetes-schema-validation

**Reply by iliapolo** [reply]
Thats not necessary because the cdk8s API is generated from the schema definitions, so a schema validation error actually surfaces as a compile time error (for typed languages only though).

**Reply by rrey-aviatrix** [reply]
That's not necessarily enough.
I just face a situtation with cdk8s (2.0.30) in golang when I was able to declare an ingress backend with both port and name which is not allowed.
The synth worked and reported no issue. I got rejected at deployment with an error saying both field ould not be provided together.

**Reply by iliapolo** [reply]
@rrey-aviatrix Can you provide the code snippet you used? I believe what you're describing is a situation where the schema defines two optional properties, because either one can be used, but not both. 

For example, the schema for a volume is:

```json
"io.k8s.api.core.v1.Volume": {
  "description": "Volume represents a named volume in a pod that may be accessed by any container in the pod.",
  "properties": {
    "awsElasticBlockStore": {
      "$ref": "#/definitions/io.k8s.api.core.v1.AWSElasticBlockStoreVolumeSource",
      "description": "AWSElasticBlockStore represents an AWS Disk resource that is attached to a kubelet's host machine and then exposed to the pod. More info: https://kubernetes.io/docs/concepts/storage/volumes#awselasticblockstore"
    },
    "azureDisk": {
      "$ref": "#/definitions/io.k8s.api.core.v1.AzureDiskVolumeSource",
      "description": "AzureDisk represents an Azure Data Disk mount on the host and bind mount to the pod."
    },
    ...
    ...
  }
}
```

You can either specify `awsElasticBlockStore`, **or** `azureDisk` - but **not both**. 

The "not both" part is not encoded in the schema so there's no way for cdk8s to know that. These types of problems is where [cdk8s-plus](https://cdk8s.io/docs/latest/plus/) really helps because it provides an API that only allows you to specify one of them. 

@eyarz Are the schema validation tools you know of able to catch these types of errors as well? or do they only validate the YAML is a legal k8s manifest?

**Reply by rrey-aviatrix** [reply]
@iliapolo yes it is the same kind of issue with the Ingress backend.

Here is a snippet
```
	var http_port float64 = 80

	k8s.NewKubeIngress(chart, jsii.String("test"), &k8s.KubeIngressProps{
		Metadata: &k8s.ObjectMeta{
			Name: jsii.String("whatever"),
		},
		Spec: &k8s.IngressSpec{
			Rules: &[]*k8s.IngressRule{
				{
					Host: jsii.String("example.com"),
					Http: &k8s.HttpIngressRuleValue{
						Paths: &[]*k8s.HttpIngressPath{
							{
								Backend: &k8s.IngressBackend{
									Service: &k8s.IngressServiceBackend{
										Name: jsii.String("some-srv"),
										Port: &k8s.ServiceBackendPort{
											Name:   jsii.String("http"),
											Number: &http_port,
										},
									},
								},
								PathType: jsii.String("Prefix"),
								Path:     jsii.String("/"),
							},
						},
					},
				},
			},
		},
	})
```
The synth is quiet and when deploying:
```
The Ingress "whatever" is invalid: spec.rules[0].http.paths[0].backend: Invalid value: "": cannot set both port name & port number
```

So having a post synthesis validation that could raise this kind of issues in the CI before any attempt to deploy would be nice.

**Reply by eyarz** [reply]
An interesting fact about K8s schema - not all validations are defined in the schema. There are some validations that you will expect to be n the schema but are actually validated by a Go script before the manifest is deployed (on the server side). For example, [validating labels value](https://stackoverflow.com/questions/61414031/invalid-label-value-on-k8s-on-gcp#:~:text=invalid%20label%20value%3A%20%22%24%7B,Za%2Dz0%2D9_.%5D) is not part of the schema definition.

Therefore, existing schema validation tools (like _Kubeconform_ & _Kubeval_) will not catch this type of misconfigurations, but Datree has a set of built-in rules (out-of-the-box) to catch them (e.g. [Ensure workload has valid label values](https://hub.datree.io/built-in-rules/ensure-labels-value-valid))

@iliapolo if the cdk8s API is generated from the schema definitions, it can't create conflicts if I'm using (in my cluster) a different K8s version?

**Reply by iliapolo** [reply]
@eyarz You mean conflicts between two different versions? 

The version cdk8s uses is fully controlled by the user - so if your cluster is 1.22.0, you should tell cdk8s to generate the API for version 1.22.0. This is an option in the cdk8s configuration file.

**Comment by iliapolo** [comment]
After some more time to think about this, summarizing what I came up with so far:

## Why

I think the need for Kubernetes manifest validation in general has already been substantiated, so in this section I want to answer the question: 

*Why should **cdk8s** in particular offer manifest validation mechanisms?*

One could argue that, since cdk8s produces native k8s manifests, users can simply invoke any validation tool they want after running cdk8s synth, and that there really is no need for explicit cdk8s support for it. While this is technically true, I think cdk8s is well positioned to make setting up this workflow much easier. Before we dive into why, I'd like to lay out an assumption we should agree on:

**“In large organizations, validation policies are defined by a central security & compliance team, while manifests are written by all development teams“**

Note that we are not focused on deploy time here, i.e preventing misconfiguration from landing in your *cluster*.
For that, we assume organizations have sufficient deployment guardrails, be it a central CD Pipeline, an [Admissions Controller](https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/), or a GitOps plugin. We are focused on authoring time, i.e preventing misconfiguration from landing in your *manifest*, thus reducing feedback time and increasing development velocity.

So, without explicit cdk8s support, how can organizations ensure their dev teams are writing compliant manifests?
An organization would have to figure out a way to inject itself into the developer workflow, which normally consists of:

### 1) Local editing

There are basically two ways of achieving this:

* Guides that instruct developers to run specific commands before pushing manifests.
* Creating wrapper tools and instruct developer to interact solely with them.

Guides are great as a source of information, but they are not a mechanism. They will most likely
fall short over time, as requirements change and dev teams grow larger and larger.

Wrapper tools are also good, but like guides, there's really no way of ensuring developers use them. They also pose an additional maintenance burden on a central team, that might not have the necessary skills to create such tools. Also, if you think about, cdk8s is already a wrapper tool for authoring manifests, and developers are already using it, so doesn't it make sense for cdk8s to offer this capability? 

Essentially, without cdk8s, the authoring of manifests is inherently decoupled from its validation, and thus will
always be prone to a lot more human error / oversight.

### 2) Pull Request

An organization can add a validation command to the CI pipelines that validate pull requests. This is not so easy to achieve, as CI pipelines can greatly vary between teams. Composing → Publishing → Reusing workflows can be fairly complicated. Also, it delays the validation by the latency of the pipeline, and introduces another failure point for the feedback loop.  

-----------------------------------------------------

To summarize, given developers are already using cdk8s to author manifests locally, it is the perfect 
place to incorporate validation mechanisms and provide quick feedback. There are few additional reasons this should be integrated with cdk8s:

* When authoring manifests with cdk8s, and especially true for cdk8s-plus, resource names are not easily backtracked to their source in the cdk8s application. A validation that isn't aware of the construct tree will probably not provide enough information for easily locating and fixing the violation.
* Some organizations are still reluctant to give their developers the freedom to author k8s manifests. This is partly because it’s hard to setup a proper validation mechanism. Providing this capability in cdk8s can help these organizations, and actually become a good reason for using cdk8s, if they are not already doing so.

**Comment by iliapolo** [comment]
## What

The proposal here is to allow cdk8s to integrate with third-party k8s manifest validation tools.
Based on user configuration, cdk8s will invoke these tools during synthesis, and optionally reject 
manifests that exhibit violations.

For example, given the following cdk8s code:

```typescript
export class MyChart extends Chart {
  constructor(scope: Construct, id: string) {
    super(scope, id);

    new k8s.KubeDeployment(this, 'Deployment', {
      spec: {
        template: {
          containers: [{ image: 'nginx', name: 'main' }]
        }
      }
    });

  }
}

const app = new App();
new MyChart(app, 'Chart');
app.synth();
```

Along with configuring* [datree](https://datree.io/) as the validation tool, running cdk8s synth should produce something similar to:

```console
❌  Ensure each container has a configured CPU request  [1 occurrence]
    - construct: Chart/Deployment (kind: Deployment)
💡  Missing property object `requests.cpu` - value should be within the accepted boundaries recommended by the organization

❌  Ensure each container image has a pinned (tag) version  [1 occurrence]
    - construct: Chart/Deployment (kind: Deployment)
💡  Incorrect value for key `image` - specify an image version to avoid unpleasant "version surprises" in the future
```

> *How exactly this configuration looks like is outlined in the “How” section.

### Requirements

#### Open System

There are many validation tools out there. We would like cdk8s to be able to support all of them, without requiring code changes in the core library for every new tool. Ideally, these integrations will be maintained by the authors of the validation tools, and not by the cdk8s team. 

#### Construct Aware

By default, the validation tools operate on k8s resources. The reports they generate usually point to specific resources by name. However, when writing cdk8s applications, providing a resource name is actually discouraged, so the report needs to point to a specific construct instead.

In addition, because of the imperative nature of cdk8s, it’s not always easy to figure out where exactly in the code was a construct even created. We would like the report to also provide a sort of traceback to help developers locate the specific code being used.

#### Distributable Config

As mentioned earlier, we are operating under the assumption that validation config (e.g which tool, what policy) is defined by central compliance and security teams, not by the developers themselves. This means that these teams need a way to publish common configuration that all developer teams can then easily consume.

**Comment by iliapolo** [comment]
## How

### Distributable Config

Validations will be declared in the cdk8s.yaml configuration file, and the cdk8s CLI will try to load and execute them. For example, a configuration will look something like:

```yaml
app: node main.js
language: typescript
imports:
  - k8s
validations:
  - http://mycompany.com/cdk8s-validations.yaml
```

Where `http://mycompany.com/cdk8s-validations.yaml` is a file published by the central team, and contains:

```yaml
- package: datree-cdk8s-plugin
  class: DatreeValidation
  config:
    policyUrl: http://mycompany.com/k8s-datree-policy.yaml
- package: checkov-cdk8s-plugin
  class: CheckovValidation
  config:
    rulesUrl: http://mycompany.com/k8s-checkov-policy.yaml
```

> I intentionally used `policyUrl` and `rulesUrl` to emphasize that the config of each validation can be different.

Every manifest generated as part of a cdk8s app that uses this config, will undergo both datree and checkov validations, based on the configuration properties provided by the central team. This way, central teams only maintain the validations config and policies, and developers only need to point their config file to the appropriate validations URL.

At this point you might be wondering - how would organizations ensure developers actually configure the validations property to point to the right location? What prevents them from simply omitting it? Well, nothing really, when developers are writing manifests, there’s always going to be some setup they need to do. Our goal is to minimize it, and make it as easy as possible. The way I see this playing out is:

1. Developer writes a cdk8s application without the correct validations config.
2. Developer pushes non-compliant manifests.
3. Deployment guardrails catch these violations, and instruct the developer to add a validations property to their config file.
4. Developer adds the validations property, and avoids these violations going forward.

_**FAQ**_

**Q**. Why not inline the validations config in the cdk8s.yaml file? Why this level of indirection? 
**A**. We are trying to minimize the amount of local setup developers need to do. Adding this level of indirection means that no matter how the configuration changes, they will always need to perform just a single change.

### Open System

For the validation mechanism to be an open system, we need to create a sort of plugin mechanism. As already alluded in the previous section, third-parties should be able to implement and publish a validation plugin, and cdk8s should be able to load and execute it.

Essentially what this means is that we need to define an interface, that third-parties can implement:

```typescript
export interface Validation {

  /**
   * Run the validation tool and produce a report.
   */
  validate(manifests: string[]): Violation[];
}
```

Where the `Violation` struct is defined as follows:

```typescript
/**
 * Represents a distinct violation in the manifest.
 */
export interface Violation {
  /**
   * The name of resource exhibiting the violation.
   */
  readonly resourceName: string;
  /**
   * The violation message.
   */
  readonly message: string;
  
  /**
   * The manifest file this violation originated in.
   */
  readonly manifestPath: string;
}
```

With these interfaces in-place, this is how a datree implementation would potentially look like:

```typescript
import { Validation, Violation } from 'cdk8s-cli';

export interface DatreeValidationProps {
  /**
   * URL for the policy to use.
   */
  readonly policyUrl: string;
}

export class DatreeValidation implements Validation {
  
  private readonly policyUrl: string;
  
  constructor(props: DatreeValidationProps) {
    this.policyUrl = props.policyUrl;
  }
  
  public validate(manifests: string[]): Violation[] {
    const report: Violation[] = [];
    for (const manifest of manifests) {
      const out = execSync(`datree test --output json ${manifest}`);
      report.push(...parseOutput(out));
    }
    return report;
  }
} 
```

During cdk8s synth, the CLI will look at the validations config property, download it from the remote source if necessary, and proceed to load and execute the plugins.

_**FAQ**_

**Q**. Who is responsible for datree CLI to exist on the system? the user? cdk8s?
**A*.* Not sure, let’s discuss.

**Q**. Should cdk8s synth fail if the validation results in violations?
**A*.* Intuitively i’d say yes, but in order to debug the violation, the manifest itself will most likely be needed. So we will need to generate it, even if the validation fails. We should probably still exist with non-zero though, to indicate an error. 

### Construct Aware

The ability to traceback to constructs world, will be enabled by adding a few annotations to the manifests generated by cdk8s. More concretely, we add cdk8s.io/construct.path and cdk8s.io/construct.traceback annotations:

```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: chart-deployment-c8a3b439
  annotations:
    cdk8s.io/construct.path: Chart/Deployment
    cdk8s.io/construct.traceback:
      - deployment.ts:25
      - main.ts:15
...
...
```

This should be fairly easily achievable by adding this logic in our base [ApiObject](https://github.com/cdk8s-team/cdk8s-core/blob/2.x/src/api-object.ts) class. Once we have it, we can find the violating resource by name, since the name is provided in the report, and then extract both the traceback and path from these annotations, to produce the final report we show to users.

**Comment by iliapolo** [comment]
## Future Possibilities

### Advanced in-memory validations

The mechanism described here suggests the validations are executed by the CLI. We could however, run the validations in the framework, as part of App.synth. This is somewhat more involved, but can allow writing more advanced plugins and validations, since they can have access to the entire construct tree.

### Organizational Defaults

Similarly to validations, organizations also want to define their own org specific defaults that should be used in manifests.
For example, one organization may want 512MB default container memory, while another may want 2GB. It would be nice to provide a mechanism by which central teams can provide these defaults. Note that these defaults may also be relevant in the context of cdk8s-plus. For example, one organization can choose to always create network isolated pods by default, and one may choose not to. These types of defaults are impossible to apply out-of-process, and will therefore probably require an in-memory integration point - most like cdk8s.App.

**Comment by iliapolo** [comment]
Following are the list of tasks to implement this capability:

- https://github.com/cdk8s-team/cdk8s-core/issues/688
- https://github.com/cdk8s-team/cdk8s-cli/issues/460
